File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.py
# 입력 처리
import sys
N=int(sys.stdin.readline().rstrip())

from collections import deque

for _ in range(N):
    length=int(sys.stdin.readline().rstrip())
    s_x,s_y=map(int,sys.stdin.readline().rstrip().split())
    e_x,e_y=map(int,sys.stdin.readline().rstrip().split())
    visited=[[0 for _ in range(length)] for _ in range(length)]

    # 오른쪽 위 대각선부터 왼쪽 위 대각선까지 8방향
    dx=[1,2,2,1,-1,-2,-2,-1]
    dy=[2,1,-1,-2,-2,-1,1,2]

    def bfs(s_x,s_y):
        q=deque()
        q.append([s_x,s_y])

        while q:
            x,y=q.popleft()

            for i in range(8):
                nx=x+dx[i]
                ny=y+dy[i]
                if nx>=0 and nx<length and ny>=0 and ny<length:
                    if visited[y][x]+1 < visited[ny][nx] or visited[ny][nx]==0:
                        visited[ny][nx]=visited[y][x]+1
                        q.append([nx,ny])
    if s_x==e_x and s_y ==e_y:
        print(0)
    else:
        bfs(s_x,s_y)
        print(visited[e_y][e_x])

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.java
import java.util.*;
import java.io.*;

class Main {
    static int[] dx={1,2,2,1,-1,-2,-2,-1};
    static int[] dy={2,1,-1,-2,-2,-1,1,2};
    static int N,answer;
    static boolean[][] visited;
    
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        int T = Integer.parseInt(br.readLine());
        
        for(int i=0;i<T;i++){
            N = Integer.parseInt(br.readLine());
            visited = new boolean[N][N];
            
            StringTokenizer st = new StringTokenizer(br.readLine());
            int s_x = Integer.parseInt(st.nextToken());
            int s_y = Integer.parseInt(st.nextToken());
            
            st = new StringTokenizer(br.readLine());
            int e_x = Integer.parseInt(st.nextToken());
            int e_y = Integer.parseInt(st.nextToken());
            
            answer=0;
            bfs(s_y,s_x,e_y,e_x);
            System.out.println(answer);
        }
    }
    
    static void bfs(int s_y, int s_x, int e_y, int e_x) {
    Deque<int[]> q = new ArrayDeque<>();
    q.add(new int[]{s_y, s_x, 0});
    visited[s_y][s_x] = true;

    while (!q.isEmpty()) {
        int[] curr = q.poll();
        int y = curr[0];
        int x = curr[1];
        int count = curr[2];

        if (y == e_y && x == e_x) {
            answer = count;
            return;
        }

        for (int i = 0; i < 8; i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];

            if (ny >= 0 && ny < N && nx >= 0 && nx < N && !visited[ny][nx]) {
                visited[ny][nx] = true;
                q.add(new int[]{ny, nx, count + 1});
                }
            }
        }
    }
}